src/java.base/share/classes/sun/security/util/DerInputStream.java

```diff
@@ -305,4 +306,90 @@ static int getDefiniteLength(InputStream in) throws IOException {
      * empty.
      */
     public int available() { return end - pos; }
+
+    /**
+     * Ensures there is no more data. This can be called when the last
+     * expected field is parsed, and we need to make sure no unread is left.
+     *
+     * @throws IOException if the end is NOT reached yet
+     */
+    public void atEnd() throws IOException {
+        if (available() != 0) {
+            throw new IOException("Extra unused bytes");
+        }
+    }
+
+    /**
+     * Checks if the tag of the next DerValue matches the rule.
+     *
+     * @param rule the rule to check for the tag.
+     * @return true if matches, false if not or stream is at end.
+     * @throws IOException if an I/O error happens while peeking the byte
+     */
+    public boolean checkNextTag(Predicate<Byte> rule) throws IOException {
+        return available() > 0 && rule.test(data[pos]);
+    }
+
+    /**
+     * Detect if the tag of the next DerValue is the specified one.
+     *
+     * @param tag the expected tag
+     * @return true if matches, false if not or stream is at end.
+     * @throws IOException if an I/O error happens while peeking the byte
+     */
+    public boolean checkNextTag(byte tag) throws IOException {
+        return checkNextTag(t -> t == tag);
+    }
+
+    /**
+     * Detect if the next DerValue is a context-specific value
+     * tagged by {@code n}.
+     *
+     * @param n the expected tag
+     * @return true if matches, false if not or stream is at end.
+     * @throws IOException if an I/O error happens while peeking the byte
+     */
+    public boolean seeOptionalContextSpecific(int n) throws IOException {
+        return checkNextTag(t -> (t & 0x0c0) == 0x080 && (t & 0x01f) == n);
```

Comment:
Minor nit for clarity, the 0x080 could use the static DerValue.TAG_CONTEXT.

On an unrelated note, it's a bummer that we don't have names for the typical masking constants 0x0c0 and 0x01f for getting at the tag class and number values (or static methods maybe).  But those numeric literals are only in wide use in DerValue.java and maybe not worth changing since you don't touch that file.  It might make that code a bit clearer, but it is just a nit.

Reply:
In fact, I tried using DerValue.TAG_CONTEXT at the beginning but it's defined as a byte so it's actually -128, and will never equal to any `t & 0x0c0`. Even the `DerValue::isContextSpecific` uses 0x80 directly. (Unfortunately this is not a static method otherwise I'll use it).

Reply:
Yep, you're right.  Sign extension strikes again!  Nevermind then, it's fine as-is.